plugins/pixart-rf/fu-pxi-device.c

```diff
@@ -0,0 +1,515 @@
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_READ,
+			     "checksum fail, got %x, expected %x",
+			     checksum_tmp,
+			     checksum);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_pxi_device_reset (FuPxiDevice *self, gsize fw_sz, GError **error)
+{
+	guint8 req[FU_PXI_DEVICE_OTA_BUF_SZ] = {
+		PXI_HID_DEV_OTA_OUTPUT_REPORT_ID,
+		FU_PXI_DEVICE_CMD_FW_MCU_RESET,
+		fw_sz, /* FIXME: this is size_t being sqeezed into uint8_t?! */
+	};
+	fu_device_set_status (FU_DEVICE (self), FWUPD_STATUS_DEVICE_RESTART);
+	if (!fu_udev_device_pwrite_full (FU_UDEV_DEVICE (self), 0,
+					 req, sizeof(req),
+					 error)) {
+		g_prefix_error (error, "failed to reset: ");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_pxi_device_fw_ota_init (FuPxiDevice *self, GError **error)
+{
+	const guint8 req[] = {
+		PXI_HID_DEV_OTA_OUTPUT_REPORT_ID,
+		FU_PXI_DEVICE_CMD_FW_OTA_INIT,
+	};
+	return fu_udev_device_pwrite_full (FU_UDEV_DEVICE (self), 0,
+					   req, sizeof(req), error);
+}
+
+static gboolean
+fu_pxi_device_fw_ota_init_new (FuPxiDevice *self, gsize fw_sz, GError **error)
+{
+	guint8 res[FU_PXI_DEVICE_OTA_BUF_SZ];
+	guint8 fw_version[10] = { 0x0 };
+	g_autoptr(GByteArray) req = g_byte_array_new ();
+
+	/* write fw ota init new command */
+	fu_byte_array_append_uint8 (req, PXI_HID_DEV_OTA_OUTPUT_REPORT_ID);
+	fu_byte_array_append_uint8 (req, FU_PXI_DEVICE_CMD_FW_OTA_INIT_NEW);
+	fu_byte_array_append_uint32 (req, fw_sz, G_LITTLE_ENDIAN);
+	fu_byte_array_append_uint8 (req, 0x0);	/* OTA setting */
+	g_byte_array_append (req, fw_version, sizeof(fw_version));
+	if (!fu_udev_device_pwrite_full (FU_UDEV_DEVICE (self), 0x0,
+					 req->data, req->len, error))
+		return FALSE;
+
+	/* delay for read command */
+	g_usleep (30 * 1000);
+
+	/* read fw ota init new command */
+	res[0] = PXI_HID_DEV_OTA_FEATURE_REPORT_ID;
+	res[1] = FU_PXI_DEVICE_CMD_FW_OTA_INIT_NEW;
+	if (!fu_pxi_device_get_feature (self, res, sizeof(res), error))
+		return FALSE;
+
+	/* shared state */
+	if (!fu_common_read_uint8_safe (res, sizeof(res), 0x2,
+					&self->status, error))
+		return FALSE;
+	if (!fu_common_read_uint8_safe (res, sizeof(res), 0x3,
+					&self->new_flow, error))
+		return FALSE;
+	if (!fu_common_read_uint16_safe (res, sizeof(res), 0x4,
+					 &self->offset, G_LITTLE_ENDIAN, error))
+		return FALSE;
+	if (!fu_common_read_uint16_safe (res, sizeof(res), 0x6,
+					 &self->checksum, G_LITTLE_ENDIAN, error))
+		return FALSE;
+	if (!fu_common_read_uint32_safe (res, sizeof(res), 0x8,
+					 &self->max_object_size, G_LITTLE_ENDIAN, error))
+		return FALSE;
+	if (!fu_common_read_uint16_safe (res, sizeof(res), 0xc,
+					 &self->mtu_size, G_LITTLE_ENDIAN, error))
+		return FALSE;
+	if (!fu_common_read_uint16_safe (res, sizeof(res), 0xe,
+					 &self->prn_threshold, G_LITTLE_ENDIAN, error))
+		return FALSE;
+	if (!fu_common_read_uint8_safe (res, sizeof(res), 0x10,
+					&self->spec_check_result, error))
+		return FALSE;
```

Comment:
@jyu99 is this offset correct?

Reply:
Could I use below to replace append
```
struct _FuPxiDevice {
	FuUdevDevice	 parent_instance;
	ret_fw_ota_init_new_cmd	ret_fw_ota_init_new;
};
```
```
	if (!fu_memcpy_safe ((guint8 *) &self->ret_fw_ota_init_new, sizeof(ret_fw_ota_init_new_cmd), 0,		/* dst */
			     (guint8 *)  &evt.evt_param.ret_param.fw_ota_init_new, sizeof(ret_fw_ota_init_new_cmd), 0,			/* src */
			     sizeof(ret_fw_ota_init_new_cmd), error))
```

Reply:
No, because all the values like max_object_size are going to be wrong on big endian systems.